docs/howtos/extracting_intermediates.rst

```diff
@@ -0,0 +1,272 @@
+can define all submodules in ``setup`` and avoid using ``nn.compact`` altogether.
+
+.. testcode::
+
+  class RefactoredCNN(nn.Module):
+    def setup(self):
+      self.features = Features()
+      self.classifier = Classifier()
+
+    def __call__(self, x):
+      x = self.features(x)
+      x = self.classifier(x)
+      return x
+
+  class Features(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Conv(features=32, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = nn.Conv(features=64, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = x.reshape((x.shape[0], -1))  # flatten
+      return x
+
+  class Classifier(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(features=256)(x)
+      x = nn.relu(x)
+      x = nn.Dense(features=10)(x)
+      x = nn.log_softmax(x)
+      return x
+
+  @jax.jit
+  def init(key, x):
+    variables = RefactoredCNN().init(key, x)
+    return variables['params']
+
+  @jax.jit
+  def features(params, x):
+    return RefactoredCNN().apply({"params": params}, x,
+      method=lambda module, x: module.features(x))
+
+  params = init(jax.random.PRNGKey(0), batch)
+
+  features(params, batch)
+
+
+Use `capture_intermediates`
+---------------------------
+
+Linen supports the capture of intermediate return values from submodules automatically without any code changes.
+This pattern should be considered the "sledge hammer" approach to capturing intermediates.
+As a debugging and inspection tool it is very useful but using the other patterns described in this howto.
+
+In the following code example we check if any intermediate activations are non-finite (NaN or infinite):
+
+.. testcode::
+
+  @jax.jit
+  def init(key, x):
+    variables = CNN().init(key, x)
+    return variables
+
+  @jax.jit
+  def predict(variables, x):
+    y, state = CNN().apply(variables, x, capture_intermediates=True, mutable=["intermediates"])
+    intermediates = state['intermediates']
+    fin = jax.tree_map(lambda xs: jnp.all(jnp.isfinite(xs)), intermediates)
+    return y, fin
+
+  variables = init(jax.random.PRNGKey(0), batch)
+  y, is_finite = predict(variables, batch)
+  all_finite = all(jax.tree_leaves(is_finite))
+  assert all_finite, "non finite intermediate detected!"
+
+
+Use ``nn.Sequential``
+---------------------
+
+You could also define ``CNN`` using a simple implementation of a ``Sequential`` combinator (this is quite common in more stateful approaches). This may be useful
+for very simple models and gives you arbitrary model
+surgery. But it can be very limiting -- if you even want to add one conditional, you are 
+forced to refactor away from ``nn.Sequential`` and structure
+your model more explicitly.
+
+.. testcode::
+
+  class Sequential(nn.Module):
+    submodules: Sequence[nn.Module]
+
+    def setup(self):
```

Comment:
I think this `setup` block is no longer necessary, after a recent change by @levskaya

Reply:
True